src/store/SwapMetaView.cc

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright (C) 1996-2022 The Squid Software Foundation and contributors
+ *
+ * Squid software is distributed under GPLv2+ license and includes
+ * contributions from numerous individuals and organizations.
+ * Please see the COPYING and CONTRIBUTORS files for details.
+ */
+
+#include "squid.h"
+#include "base/Raw.h"
+#include "base/TextException.h"
+#include "debug/Stream.h"
+#include "sbuf/Stream.h"
+#include "store/SwapMetaView.h"
+
+#include <iostream>
+
+namespace Store {
+
+/// properly reports or rejects a problematic raw swap meta field type
+static void
+HandleBadRawType(const RawSwapMetaType type)
+{
+    if (ReservedSwapMetaType(type)) {
+        debugs(20, 3, "ignoring swap meta field with a reserved type: " << int(type));
+        return;
+    }
+
+    if (DeprecatedSwapMetaType(type)) {
+        debugs(20, DBG_CRITICAL, "ERROR: Ignoring swap meta field with a deprecated type: " << int(type));
+        return;
+    }
+
+    // TODO: Instead of assuming that all future swap meta types can be ignored
+    // (and some should be reported at level-0/1), future-proof this code by
+    // using a type bit to define whether to silently ignore a swap meta field
+    // with that type (or even the whole Store entry with that field).
+
+    const auto typeValuesWeMightAdd = 10;
+    // compute "type > RawSwapMetaTypeTop() + typeValuesWeMightAdd" w/o overflow
+    if (type >= typeValuesWeMightAdd && type - typeValuesWeMightAdd > RawSwapMetaTypeTop()) {
+        debugs(20, DBG_CRITICAL, "ERROR: Malformed cache storage; ignoring swap meta field with an unexpected type: " << int(type));
+        return;
+    }
+
+    if (type > RawSwapMetaTypeTop()) {
+        debugs(20, 3, "ignoring swap meta field with a presumed future type: " << int(type));
+        return;
+    }
+
+    Assure(type <= RawSwapMetaTypeBottom);
+    debugs(20, DBG_CRITICAL, "ERROR: Malformed cache storage; ignoring swap meta field with an invalid type: " << int(type));
```

Comment:
> This is input validation for SwapMetaView() parser. Assure is not supposed to be used for input validation.

This Assure() statement is not about input validation: If the condition is false, we would need to fix Squid _code_. Once we get to this Assure() line/code, we state that the previous input validation handled all cases it was supposed to handle and, hence, we are left with the cases we need to handle here. We should use assert() or Assure() for that code logic check, of course. 

> Assure condition appears to be reversed to casual reader. It allows negative type codes, but throws on unknown positives.

In commit 0563d35, I added an explicit description to this Assure() in hope to help code readers.

The condition is exactly what it should be -- `type <= RawSwapMetaTypeBottom`. Yes, it "allows" types with negative values (assuming `char` is signed) and types with the RawSwapMetaTypeBottom value to reach the handling code below, as it should. Those are the values not handled by the previous code; they need special/dedicated handling in this code.

This code "throws on unknown positives" because there should be no unknown positives at this point -- the earlier code should have handled them already. If they exist at this point, it is a Squid bug. I am not sure where the confusion about this code is coming from, but I will note that the function you are reviewing is dedicated to handling _bad_ swap meta types. It is properly described as such. Good (i.e. "honored") swap meta types are handled elsewhere:

```C++
    if (HonoredSwapMetaType(rawType))
        type = static_cast<SwapMetaType>(rawType);
    else
        HandleBadRawType(rawType);
```

FWIW, the specific code change suggestion you posted would convert code logic bugs into "invalid type" input validation exceptions. We should not do that.

> the debugs should not be CRITICAL messages

Changed to DBG_IMPORTANT in hope to make progress (commit 077d9ee).

FWIW, the PR was trying to preserve the official behavior here:

```diff
-    if (type <= STORE_META_VOID ...) {
-        debugs(20, DBG_CRITICAL, "ERROR: storeSwapMetaUnpack: bad type (" << type << ")!");
```

> Squid is about to handle it in the caller.

Not really. This debugs() _is_ essentially the handling for this field type -- the high-level caller that handles supported fields does not even know about this field type. AFAICT, this PR does not change that official behavior.

Reply:
> > This is input validation for SwapMetaView() parser. Assure is not supposed to be used for input validation.
> 
> This Assure() statement is not about input validation: 
...
> 
> In commit [0563d35](https://github.com/squid-cache/squid/commit/0563d3516ab9208d86c1c7da0a777bd77f42ae45), I added an explicit description to this Assure() in hope to help code readers.

The comment you added very clearly indicates that this Assure is *definitely* about input validation - specifically it is validating the non-existence of a subset of potential bad inputs.

If you are certain that such values cannot ever occur, use an assert(). Otherwise this should nowdays throw without using Assure(). AFAICT, this is the type of "abuse" you complained about being done with Must().

Reply:
```C++
Assure2(type <= RawSwapMetaTypeBottom, "handled all other bad types above");
```

> The comment you added very clearly indicates that this Assure is definitely about input validation - specifically it is validating the non-existence of a subset of potential bad inputs.

The comment clearly does not match the above opinion. The comment explicitly says that the code has **handled** all other bad inputs above. It does not say that those inputs do not _exist_; it strongly implies the _opposite_ -- those inputs do exist (and have been handled). The assertion validates code logic (i.e. we handled something we wanted to handle), not input (i.e. whether certain input values are possible).

> If you are certain that such values cannot ever occur, use an assert().

Done in d04824b285defb327307a88e265bb0bf3e0311e1 in hope to make progress.

> this is the type of "abuse" you complained about being done with Must().

It is not. This code does not validate input. If this assertion fails, we will need to find and fix a bug in our code.